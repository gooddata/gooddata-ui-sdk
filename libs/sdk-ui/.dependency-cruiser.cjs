const depCruiser = require("../../common/config/dep-cruiser/default.config");

const options = {
    forbidden: [
        ...depCruiser.DefaultRules,
        ...depCruiser.DefaultSdkRules,
        ...depCruiser.PublicLibraryRules,
        depCruiser.isolatedSubmodule("base", "src/base"),
        depCruiser.moduleWithDependencies("executions", "src/execution", ["src/base"]),
    ],
    options: depCruiser.DefaultOptions,
};

module.exports = options;
