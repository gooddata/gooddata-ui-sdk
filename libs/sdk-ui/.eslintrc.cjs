// (C) 2020 GoodData Corporation

const { tsOverride } = require("@gooddata/eslint-config/tsOverride");

module.exports = {
    extends: ["@gooddata/eslint-config/esm-react-vitest"],
    rules: {
        "import/export": "off",
    },
    overrides: [
        tsOverride(__dirname, {
            "@typescript-eslint/no-floating-promises": "off",
            "eslint-comments/no-unused-disable": "off",
            "@typescript-eslint/no-unnecessary-type-assertion": "off",
            "@typescript-eslint/unbound-method": "off",
            "@typescript-eslint/no-unsafe-argument": "off",
            "@typescript-eslint/no-unsafe-member-access": "off",
            "@typescript-eslint/require-await": "off",
            "@typescript-eslint/no-unsafe-assignment": "off",
            "@typescript-eslint/restrict-template-expressions": "off",
            "@typescript-eslint/no-unsafe-return": "off",
            "@typescript-eslint/naming-convention": "off",
            "@typescript-eslint/no-unsafe-call": "off",
            "@typescript-eslint/no-unsafe-enum-comparison": "off",
            "@typescript-eslint/no-redundant-type-constituents": "off",
            "@typescript-eslint/prefer-promise-reject-errors": "off",
        }),
    ],
};
