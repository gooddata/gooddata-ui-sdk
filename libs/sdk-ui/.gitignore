styles/css
styles/internal/css

src/base/localization/bundles/*.ts
