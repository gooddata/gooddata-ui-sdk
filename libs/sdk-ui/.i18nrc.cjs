module.exports = {
    paths: ["./src/base/localization/bundles"],
    structure: true,
    intl: true,
    html: true,
    insightToReport: true,
    usage: true,
    debug: false,
    source: "../{sdk-ui,sdk-ui-charts,sdk-ui-pivot,sdk-ui-kit,sdk-ui-filters,sdk-ui-vis-commons,sdk-ui-geo}/src/**/*.{ts,js,tsx,jsx}",
    rules: [
        {
            pattern: [/.+/],
        },
        // some messages need to be included for the sake of older dashboard plugins
        {
            dir: /src\/base\/localization\/bundles/,
            pattern: /^(gs.filter.loading|attrf.all|attrf.all_except|attributeFilterDropdown.emptyValue)$/,
            ignore: true,
        },
        /**
         * These ignored messages are used in the attribute filter implementation for dashboard component, so we need to have these messages
         * here even though they are used within the sdk-ui-dashboard code.
         */
        {
            dir: /src\/base\/localization\/bundles/,
            pattern:
                /^(attributesDropdown.filterConfiguredMessage|attributesDropdown.noConnectionMessage|attributesDropdown.attributeNameWithData)$/,
            ignore: true,
        },
    ],
};
