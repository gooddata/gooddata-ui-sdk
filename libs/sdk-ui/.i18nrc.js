module.exports = {
    paths: ["./src/base/localization/bundles"],
    structure: true,
    intl: true,
    html: true,
    insightToReport: true,
    usage: true,
    debug: false,
    source: "../{sdk-ui,sdk-ui-charts,sdk-ui-pivot,sdk-ui-kit,sdk-ui-filters,sdk-ui-vis-commons,sdk-ui-geo}/src/**/*.{ts,js,tsx,jsx}",
    rules: [
        {
            // some messages need to be included for the sake of older dashboard plugins
            pattern:
                /^(?!gs.filter.loading|attrf.all|attrf.all_except|attributeFilterDropdown.emptyValue).*$/,
            filterTranslationFile: true,
        },
    ],
};
