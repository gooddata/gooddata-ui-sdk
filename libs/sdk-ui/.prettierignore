#we ignore these files because they are generated after build
src/base/localization/bundles/*.ts
