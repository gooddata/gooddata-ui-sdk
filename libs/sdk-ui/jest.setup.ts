// (C) 2019 GoodData Corporation
/* eslint-disable @typescript-eslint/no-var-requires */
import "jest-enzyme";
import * as raf from "raf";
import "@testing-library/jest-dom/extend-expect";

const enzyme = require("enzyme");
const Adapter = require("@wojtekmaj/enzyme-adapter-react-17");

enzyme.configure({ adapter: new Adapter() });

raf.polyfill();
