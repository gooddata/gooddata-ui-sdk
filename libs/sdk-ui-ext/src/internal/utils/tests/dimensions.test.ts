// (C) 2007-2020 GoodData Corporation
import cloneDeep from "lodash/cloneDeep.js";
import { VisualizationTypes, BucketNames } from "@gooddata/sdk-ui";
import {
    generateDimensions,
    generateStackedDimensions,
    getHeadlinesDimensions,
    getPivotTableDimensions,
} from "../dimensions.js";
import {
    bucketsFind,
    IBucket,
    IInsightDefinition,
    insightBucket,
    insightBuckets,
    ITotal,
    newBucket,
    insightSetProperties,
    newTotal,
    newMeasure,
    uriRef,
    newAttribute,
} from "@gooddata/sdk-model";
import { ReferenceRecordings } from "@gooddata/reference-workspace";
import { describe, it, expect } from "vitest";

const singleMeasureInsight = ReferenceRecordings.Insights.Headline.SingleMeasure.obj as IInsightDefinition;
const singleAttributeInsight = ReferenceRecordings.Insights.PivotTable.SingleAttribute
    .obj as IInsightDefinition;
const insightWithProperties = insightSetProperties(singleMeasureInsight, {
    controls: { xaxis: { visible: true } },
});

/*
 * WARNING: tests in this file rely on mutating insight and mutating buckets. We should revisit all the tests here.
 * Ideally, all these _component_ tests (they test the dimension generator) should use recorded insights; no mutations
 * here. Even further, all these tests should be replaced by a single massive parameterized test taking many
 * different recorded insights and verifying dimensionality.
 */

function getVisualizationBucket(newVis: IInsightDefinition, bucketName: string): IBucket {
    const buckets = insightBuckets(newVis);
    let bucket = bucketsFind(buckets, bucketName);

    if (!bucket) {
        bucket = newBucket(bucketName);
        buckets.push(bucket);
    }

    return bucket;
}

function addMeasure(visualization: IInsightDefinition, index: number): IInsightDefinition {
    const newVis = cloneDeep(visualization);
    const measure = newMeasure(uriRef(`/gdc/md/myproject/obj/150${index}`), (m) =>
        m.localId(`m${index}`).title(`# Users Opened AD ${index}`),
    );

    const bucket = insightBucket(newVis);
    bucket.items.push(measure);

    return newVis;
}

function addAttribute(
    visualization: IInsightDefinition,
    index: number,
    bucketName: string,
): IInsightDefinition {
    const newVis = cloneDeep(visualization);
    const attribute = newAttribute(uriRef(`/gdc/md/myproject/obj/400${index}`), (a) =>
        a.localId(`a${index}`),
    );

    const bucket = getVisualizationBucket(newVis, bucketName);
    bucket.items.push(attribute);

    return newVis;
}

function addTotals(
    visualization: IInsightDefinition,
    bucketName: string,
    newTotals: ITotal[],
): IInsightDefinition {
    const newVis = cloneDeep(visualization);

    const bucket = getVisualizationBucket(newVis, bucketName);

    if (!bucket.totals) {
        bucket.totals = [];
    }

    newTotals.forEach((total) => {
        bucket.totals.push(total);
    });

    return newVis;
}

function newInsight(buckets: IBucket[]): IInsightDefinition {
    return {
        insight: {
            title: "no name",
            visualizationUrl: "classId",
            buckets,
            filters: [],
            sorts: [],
            properties: {},
        },
    };
}

describe("getHeadlinesDimensions", () => {
    it("should always return just one dimension with a measureGroup", () => {
        expect(getHeadlinesDimensions()).toMatchSnapshot();
    });
});

describe("getPivotTableDimensions", () => {
    it("should return row attributes in the first dimensions, column attributes and measureGroup in second dimension", () => {
        const buckets = [
            newBucket(
                BucketNames.MEASURES,
                newMeasure(uriRef("/gdc/md/myproject/obj/m1"), (m) =>
                    m.localId("m1").title("# Accounts with AD Query"),
                ),
            ),
            newBucket(
                BucketNames.ATTRIBUTE,
                newAttribute(uriRef("/gdc/md/myproject/obj/a1"), (a) => a.localId("a1")),
            ),
            newBucket(
                BucketNames.COLUMNS,
                newAttribute(uriRef("/gdc/md/myproject/obj/a2"), (a) => a.localId("a2")),
            ),
        ];

        expect(getPivotTableDimensions(newInsight(buckets))).toMatchSnapshot();
    });
});

describe("generateDimensions", () => {
    describe("column/bar chart", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.COLUMN)).toMatchSnapshot();
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.BAR)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.VIEW);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.COLUMN),
            ).toMatchSnapshot();
            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.BAR),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithStackAttribute = addAttribute(visualization, 1, BucketNames.STACK);

            expect(
                generateDimensions(visualizationWithStackAttribute, VisualizationTypes.COLUMN),
            ).toMatchSnapshot();
            expect(
                generateDimensions(visualizationWithStackAttribute, VisualizationTypes.BAR),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure, view attribute and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAndStackAttribute = addAttribute(
                addAttribute(visualization, 1, BucketNames.VIEW),
                2,
                BucketNames.STACK,
            );

            expect(
                generateDimensions(visualizationWithViewAndStackAttribute, VisualizationTypes.COLUMN),
            ).toMatchSnapshot();
            expect(
                generateDimensions(visualizationWithViewAndStackAttribute, VisualizationTypes.BAR),
            ).toMatchSnapshot();
        });
    });
    describe("area chart", () => {
        it("should generate dimensions for area with one measure", () => {
            const dimensions = generateDimensions(singleMeasureInsight, VisualizationTypes.AREA);
            expect(dimensions).toMatchSnapshot();
        });

        it("should generate dimensions for area with one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.VIEW);
            const dimensions = generateDimensions(visualizationWithViewAttribute, VisualizationTypes.AREA);

            expect(dimensions).toMatchSnapshot();
        });

        it("should generate dimensions for area with one measure and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithStackAttribute = addAttribute(visualization, 1, BucketNames.STACK);
            const dimensions = generateDimensions(visualizationWithStackAttribute, VisualizationTypes.AREA);

            expect(dimensions).toMatchSnapshot();
        });

        it("should generate dimensions for area with one measure, view attribute and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAndStackAttribute = addAttribute(
                addAttribute(visualization, 1, BucketNames.VIEW),
                2,
                BucketNames.STACK,
            );
            const dimensions = generateDimensions(
                visualizationWithViewAndStackAttribute,
                VisualizationTypes.AREA,
            );

            expect(dimensions).toMatchSnapshot();
        });

        it("should generate dimensions for area with one measure, two view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithTwoViewAttributes = addAttribute(
                addAttribute(visualization, 1, BucketNames.VIEW),
                2,
                BucketNames.VIEW,
            );
            const dimensions = generateDimensions(
                visualizationWithTwoViewAttributes,
                VisualizationTypes.AREA,
            );

            expect(dimensions).toMatchSnapshot();
        });
    });
    describe("heatmap", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.HEATMAP)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.VIEW);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.HEATMAP),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithStackAttribute = addAttribute(visualization, 1, BucketNames.STACK);

            expect(
                generateDimensions(visualizationWithStackAttribute, VisualizationTypes.HEATMAP),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure, view attribute and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAndStackAttribute = addAttribute(
                addAttribute(visualization, 1, BucketNames.VIEW),
                2,
                BucketNames.STACK,
            );

            expect(
                generateDimensions(visualizationWithViewAndStackAttribute, VisualizationTypes.HEATMAP),
            ).toMatchSnapshot();
        });
    });
    describe("line chart", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.LINE)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.TREND);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.LINE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.SEGMENT);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.LINE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure, view attribute and stack attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAndStackAttribute = addAttribute(
                addAttribute(visualization, 1, BucketNames.TREND),
                2,
                BucketNames.SEGMENT,
            );

            expect(
                generateDimensions(visualizationWithViewAndStackAttribute, VisualizationTypes.LINE),
            ).toMatchSnapshot();
        });
    });
    describe("pie chart", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.PIE)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.VIEW);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.PIE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and 2 view attributes", () => {
            const visualization = singleMeasureInsight;
            const visualizationWith2ViewAttributes = addAttribute(
                addAttribute(visualization, 1, BucketNames.VIEW),
                2,
                BucketNames.VIEW,
            );

            expect(
                generateDimensions(visualizationWith2ViewAttributes, VisualizationTypes.PIE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for two measures", () => {
            const visualization = singleMeasureInsight;
            const visualizationWith2Measures = addMeasure(visualization, 2);

            expect(generateDimensions(visualizationWith2Measures, VisualizationTypes.PIE)).toMatchSnapshot();
        });
    });
    describe("treemap", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.TREEMAP)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and view attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithViewAttribute = addAttribute(visualization, 1, BucketNames.VIEW);

            expect(
                generateDimensions(visualizationWithViewAttribute, VisualizationTypes.TREEMAP),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for two measures", () => {
            const visualization = singleMeasureInsight;
            const visualizationWith2Measures = addMeasure(visualization, 2);

            expect(
                generateDimensions(visualizationWith2Measures, VisualizationTypes.TREEMAP),
            ).toMatchSnapshot();
        });
    });
    describe("table", () => {
        it("should generate dimensions for one measure", () => {
            expect(generateDimensions(singleMeasureInsight, VisualizationTypes.TABLE)).toMatchSnapshot();
        });

        it("should generate dimensions for one attribute", () => {
            expect(generateDimensions(singleAttributeInsight, VisualizationTypes.TABLE)).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and attribute", () => {
            const visualization = singleMeasureInsight;
            const visualizationWithAttribute = addAttribute(visualization, 1, BucketNames.ATTRIBUTE);

            expect(
                generateDimensions(visualizationWithAttribute, VisualizationTypes.TABLE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for one measure and 2 attributes", () => {
            const visualization = singleMeasureInsight;
            const visualizationWith2Attributes = addAttribute(
                addAttribute(visualization, 1, BucketNames.ATTRIBUTE),
                2,
                BucketNames.ATTRIBUTE,
            );

            expect(
                generateDimensions(visualizationWith2Attributes, VisualizationTypes.TABLE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions for two measures", () => {
            const visualization = singleMeasureInsight;
            const visualizationWith2Measures = addMeasure(visualization, 2);

            expect(
                generateDimensions(visualizationWith2Measures, VisualizationTypes.TABLE),
            ).toMatchSnapshot();
        });

        it("should generate dimensions with totals", () => {
            const visualization = insightWithProperties;
            const visualizationWithTotals = addTotals(visualization, BucketNames.ATTRIBUTE, [
                newTotal("sum", "m1", "a1", "Sum"),
                newTotal("sum", "m2", "a1"),
                newTotal("nat", "m1", "a1"),
            ]);

            expect(generateDimensions(visualizationWithTotals, VisualizationTypes.TABLE)).toMatchSnapshot();
        });
    });
});

describe("generateStackedDimensions", () => {
    it("measure and stack by only", () => {
        const buckets = [
            newBucket(
                BucketNames.MEASURES,
                newMeasure(uriRef("/gdc/md/myproject/obj/m1"), (m) =>
                    m.localId("m1").title("# Accounts with AD Query"),
                ),
            ),
            newBucket(
                BucketNames.STACK,
                newAttribute(uriRef("/gdc/md/myproject/obj/a2"), (a) => a.localId("a2")),
            ),
        ];

        expect(generateStackedDimensions(newInsight(buckets))).toMatchSnapshot();
    });

    it("should return 2 attributes along with measureGroup", () => {
        const buckets = [
            newBucket(
                BucketNames.MEASURES,
                newMeasure(uriRef("/gdc/md/storybook/obj/m1"), (m) => m.localId("m1")),
            ),
            newBucket(
                BucketNames.ATTRIBUTE,
                newAttribute(uriRef("/gdc/md/storybook/obj/1.df"), (a) => a.localId("a1")),
                newAttribute(uriRef("/gdc/md/storybook/obj/2.df"), (a) => a.localId("a2")),
            ),
            newBucket(BucketNames.STACK),
        ];

        expect(generateStackedDimensions(newInsight(buckets))).toMatchSnapshot();
    });

    it("should return 2 attributes along with measureGroup and return 1 stack attribute", () => {
        const buckets = [
            newBucket(
                BucketNames.MEASURES,
                newMeasure(uriRef("/gdc/md/storybook/obj/m1"), (m) => m.localId("m1")),
            ),
            newBucket(
                BucketNames.ATTRIBUTE,
                newAttribute(uriRef("/gdc/md/storybook/obj/1.df"), (a) => a.localId("a1")),
                newAttribute(uriRef("/gdc/md/storybook/obj/3.df"), (a) => a.localId("a3")),
            ),
            newBucket(
                BucketNames.STACK,
                newAttribute(uriRef("/gdc/md/storybook/obj/2.df"), (a) => a.localId("a2")),
            ),
        ];

        expect(generateStackedDimensions(newInsight(buckets))).toMatchSnapshot();
    });
});
