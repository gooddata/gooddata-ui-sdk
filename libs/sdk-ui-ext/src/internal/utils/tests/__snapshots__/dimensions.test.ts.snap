// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateDimensions > area chart > should generate dimensions for area with one measure 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`generateDimensions > area chart > should generate dimensions for area with one measure and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > area chart > should generate dimensions for area with one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > area chart > should generate dimensions for area with one measure, two view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > area chart > should generate dimensions for area with one measure, view attribute and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure 2`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure and stack attribute 2`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure and view attribute 2`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure, view attribute and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > column/bar chart > should generate dimensions for one measure, view attribute and stack attribute 2`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > heatmap > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > heatmap > should generate dimensions for one measure and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > heatmap > should generate dimensions for one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > heatmap > should generate dimensions for one measure, view attribute and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "a2",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > line chart > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`generateDimensions > line chart > should generate dimensions for one measure and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > line chart > should generate dimensions for one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > line chart > should generate dimensions for one measure, view attribute and stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > pie chart > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > pie chart > should generate dimensions for one measure and 2 view attributes 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "a2",
    ],
  },
]
`;

exports[`generateDimensions > pie chart > should generate dimensions for one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > pie chart > should generate dimensions for two measures 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions for one attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a_label.f_product.product.name",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions for one measure and 2 attributes 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions for one measure and attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions for two measures 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > table > should generate dimensions with totals 1`] = `
[
  {
    "itemIdentifiers": [],
    "totals": [
      {
        "alias": "Sum",
        "attributeIdentifier": "a1",
        "measureIdentifier": "m1",
        "type": "sum",
      },
      {
        "attributeIdentifier": "a1",
        "measureIdentifier": "m2",
        "type": "sum",
      },
      {
        "attributeIdentifier": "a1",
        "measureIdentifier": "m1",
        "type": "nat",
      },
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > treemap > should generate dimensions for one measure 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateDimensions > treemap > should generate dimensions for one measure and view attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
]
`;

exports[`generateDimensions > treemap > should generate dimensions for two measures 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateStackedDimensions > measure and stack by only 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`generateStackedDimensions > should return 2 attributes along with measureGroup 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "a1",
      "a2",
      "measureGroup",
    ],
  },
]
`;

exports[`generateStackedDimensions > should return 2 attributes along with measureGroup and return 1 stack attribute 1`] = `
[
  {
    "itemIdentifiers": [
      "a2",
    ],
  },
  {
    "itemIdentifiers": [
      "a1",
      "a3",
      "measureGroup",
    ],
  },
]
`;

exports[`getHeadlinesDimensions > should always return just one dimension with a measureGroup 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`getPivotTableDimensions > should return row attributes in the first dimensions, column attributes and measureGroup in second dimension 1`] = `
[
  {
    "itemIdentifiers": [
      "a1",
    ],
  },
  {
    "itemIdentifiers": [
      "a2",
      "measureGroup",
    ],
  },
]
`;
